src/lang/cpp/internal/TypeResolver.java

```diff
@@ -430,70 +468,79 @@ private IConstructor resolveICPPBasicType(ICPPBasicType type) {
+			return resolveICPPDeferredClassInstance((ICPPDeferredClassInstance) type, origin);
+		}
+		if (type instanceof ICPPUnknownMemberClass) {
+			return resolveICPPUnknownMemberClass((ICPPUnknownMemberClass) type, origin);
+		}
+		if (type instanceof IPDOMCPPClassType) {
+			// TODO: JV, Rodin, there was nothing here? 
 			;
-		if (type instanceof ICPPClassType)
-			return resolveICPPClassType((ICPPClassType) type);
-		if (type instanceof CStructure)
-			return resolveCStructure((CStructure) type);
-		throw new RuntimeException("NYI: resolveICompositeType");
+		}
+		if (type instanceof ICPPClassType) {
+			return resolveICPPClassType((ICPPClassType) type, origin);
+		}
+		if (type instanceof CStructure) {
+			return resolveCStructure((CStructure) type, origin);
+		}
+
+		throw new RuntimeException("NYI: resolveICompositeType " + type + " @ " + origin);
 	}
 
-	private IConstructor resolveCStructure(CStructure type) {
-		return builder.TypeSymbol_struct(getDecl(type));
+	private IConstructor resolveCStructure(CStructure type, ISourceLocation origin) {
+		return builder.TypeSymbol_struct(getDecl(type, origin));
 	}
 
-	private IConstructor resolveICPPClassType(ICPPClassType type) {
+	private IConstructor resolveICPPClassType(ICPPClassType type, ISourceLocation origin) {
 		switch (type.getKey()) {
 		case ICPPClassType.k_struct:
-			return builder.TypeSymbol_struct(getDecl(type));
+			return builder.TypeSymbol_struct(getDecl(type, origin));
 		case ICPPClassType.k_union:
-			return builder.TypeSymbol_union(getDecl(type));
+			return builder.TypeSymbol_union(getDecl(type, origin));
 		case ICPPClassType.k_class:
-			return builder.TypeSymbol_class(getDecl(type));
+			return builder.TypeSymbol_class(getDecl(type, origin));
 		default:
-			throw new RuntimeException("Unknown ICompositeType key " + type.getKey());
+			throw new RuntimeException("Unknown ICompositeType key " + type.getKey() + " @ " + origin);
 		}
 	}
 
-	private IConstructor resolveICPPClassSpecialization(ICPPClassSpecialization type) {
-		ISourceLocation decl = getDecl(type.getSpecializedBinding());
+	private IConstructor resolveICPPClassSpecialization(ICPPClassSpecialization type, ISourceLocation origin) {
+		ISourceLocation decl = getDecl(type.getSpecializedBinding(), origin);
 		IListWriter templateParameters = vf.listWriter();
 		ICPPTemplateParameterMap parameterMap = type.getTemplateParameterMap();
 		Stream.of(parameterMap.getAllParameterPositions()).forEach(it -> {
 			ICPPTemplateArgument arg = parameterMap.getArgument(it);
 			if (arg != null) {
 				if (arg.isTypeValue())
-					templateParameters.append(resolveType(arg.getTypeValue()));
+					templateParameters.append(resolveType(arg.getTypeValue(), origin));
 				else
-					templateParameters.append(resolveType(arg.getNonTypeEvaluation().getType()));
+					templateParameters.append(resolveType(arg.getNonTypeEvaluation().getType(), origin));
 			} else {
-
+				// TODO: Rodin, why is this empty?
```

Comment:
Seems like a missing implementation. I don't know if the `!= null` check is relevant at all

Reply:
ok